From a9b129511b4dcb24ee3fa7458d61efefdb94c5a8 Mon Sep 17 00:00:00 2001
From: Rajib Mahapatra <Rajib.Mahapatra@amd.com>
Date: Tue, 19 May 2020 19:30:32 +0530
Subject: [PATCH] radeonsi: don't expose 16xAA on chips with 1 RB due to an
 occlusion query issue The patch  applied to MESA-19.3.5

From f80d653d701f51f00f88601707747554c9a7af1c Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marek=20Ol=C5=A1=C3=A1k?= <marek.olsak@amd.com>
Date: Thu, 14 May 2020 18:47:36 -0400
Subject: [PATCH] radeonsi: don't expose 16xAA on chips with 1 RB due to an
 occlusion query issue

Only Stoney and Raven2 are affected.

Cc: 20.0 20.1 <mesa-stable@lists.freedesktop.org>
Reviewed-by: Bas Nieuwenhuizen <bas@basnieuwenhuizen.nl>
Part-of: <https://gitlab.freedesktop.org/mesa/mesa/-/merge_requests/5047>
---
 src/gallium/drivers/radeonsi/si_state.c | 20 ++++++++++----------
 1 file changed, 10 insertions(+), 10 deletions(-)

diff --git a/src/gallium/drivers/radeonsi/si_state.c b/src/gallium/drivers/radeonsi/si_state.c
index 311453dae0a..eca54c9d1b5 100644
--- a/src/gallium/drivers/radeonsi/si_state.c
+++ b/src/gallium/drivers/radeonsi/si_state.c
@@ -1448,12 +1448,6 @@ static void si_emit_db_render_state(struct si_context *sctx)
 		if (sctx->chip_class >= GFX7) {
 			unsigned log_sample_rate = sctx->framebuffer.log_samples;
 
-			/* Stoney doesn't increment occlusion query counters
-			 * if the sample rate is 16x. Use 8x sample rate instead.
-			 */
-			if (sctx->family == CHIP_STONEY)
-				log_sample_rate = MIN2(log_sample_rate, 3);
-
 			db_count_control =
 				S_028004_PERFECT_ZPASS_COUNTS(perfect) |
 				S_028004_DISABLE_CONSERVATIVE_ZPASS_COUNTS(gfx10_perfect) |
@@ -2257,20 +2251,26 @@ static bool si_is_format_supported(struct pipe_screen *screen,
 		    !util_is_power_of_two_or_zero(storage_sample_count))
 			return false;
 
+                /* Chips with 1 RB don't increment occlusion queries at 16x MSAA sample rate,
+                 * so don't expose 16 samples there.
+                 */
+                const unsigned max_eqaa_samples = sscreen->info.num_render_backends == 1 ? 8 : 16;
+                const unsigned max_samples = 8;
+
 		/* MSAA support without framebuffer attachments. */
-		if (format == PIPE_FORMAT_NONE && sample_count <= 16)
+		if (format == PIPE_FORMAT_NONE && sample_count <= max_eqaa_samples)
 			return true;
 
 		if (!sscreen->info.has_eqaa_surface_allocator ||
 		    util_format_is_depth_or_stencil(format)) {
 			/* Color without EQAA or depth/stencil. */
-			if (sample_count > 8 ||
+			if (sample_count > max_samples ||
 			    sample_count != storage_sample_count)
 				return false;
 		} else {
 			/* Color with EQAA. */
-			if (sample_count > 16 ||
-			    storage_sample_count > 8)
+			if (sample_count > max_eqaa_samples ||
+			    storage_sample_count > max_samples)
 				return false;
 		}
 	}
-- 
2.17.1

